UniversalAlgebra/Homomorphisms/Setoid.lagda

```diff
@@ -0,0 +1,307 @@
+
+
+ homker-comp : swelldef 𝓥 β → (h : hom 𝑨 𝑩) → 𝑨 ∣≈ (ker ∣ h ∣)
+ homker-comp wd h f {u}{v} kuv = ∣ h ∣((f ̂ 𝑨) u)   ≡⟨ ∥ h ∥ f u ⟩
+                                     (f ̂ 𝑩)(∣ h ∣ ∘ u) ≡⟨ wd(f ̂ 𝑩)(∣ h ∣ ∘ u)(∣ h ∣ ∘ v)kuv ⟩
+                                     (f ̂ 𝑩)(∣ h ∣ ∘ v) ≡⟨ (∥ h ∥ f v)⁻¹ ⟩
+                                     ∣ h ∣((f ̂ 𝑨) v)   ∎
+
+-- NOT WORKING YET
+--  kercon : swelldef 𝓥 β → hom 𝑨 𝑩 → Con 𝑨 -- {α}{β}
+--  kercon wd h = ker ∣ h ∣ , {!!} -- mkcon (ker-IsEquivalence ∣ h ∣)(homker-comp wd {𝑩} h)
+
+--  kerquo : swelldef 𝓥 β → hom 𝑨 𝑩 → SetoidAlgebra _ _ -- (α ⊔ lsuc β) _
+--  kerquo wd h = 𝑨 ╱ (kercon wd h)
+
+
+-- ker[_⇒_]_↾_ : {α ρᵃ : Level} (𝑨 : SetoidAlgebra α ρᵃ)
+--               {β ρᵇ : Level} (𝑩 : SetoidAlgebra β ρᵇ)
+--  →            hom 𝑨 𝑩 → swelldef 𝓥 β → SetoidAlgebra _ _ --  (α ⊔ lsuc β) _
+-- ker[ 𝑨 ⇒ 𝑩 ] h ↾ wd = kerquo 𝑨 𝑩 wd h
+
+\end{code}
+
+
+
+
+
+
+#### Homomorphism Decomposition for SetoidAlgebras
+
+
+If `τ : hom 𝑨 𝑩`, `ν : hom 𝑨 𝑪`, `ν` is surjective, and `ker ν ⊆ ker τ`, then there exists `φ : hom 𝑪 𝑩` such that `τ = φ ∘ ν` so the following diagram commutes:
+
+```
+𝑨 --- ν ->> 𝑪
+ \         .
+  \       .
+   τ     φ
+    \   .
+     \ .
+      V
+      𝑩
+```
+
+\begin{code}
+
+module _ {α ρᵃ : Level} {𝑨 : SetoidAlgebra α ρᵃ}
+         {β ρᵇ : Level} (𝑩 : SetoidAlgebra β ρᵇ)
+         {γ ρᶜ : Level} {𝑪 : SetoidAlgebra γ ρᶜ} where
+
+ private
+  A = 𝕌[ 𝑨 ]
+  B = 𝕌[ 𝑩 ]
+  C = 𝕌[ 𝑪 ]
+
+ open import Axiom.Extensionality.Propositional    using    ()
+                                                  renaming (Extensionality to funext)
+
+ HomFactor : funext α β → swelldef 𝓥 γ
+  →          (τ : hom 𝑨 𝑩)(ν : hom 𝑨 𝑪)
+  →          kernel ∣ ν ∣ ⊆ kernel ∣ τ ∣ → IsSurjective ∣ ν ∣
+             --------------------------------------------------
+  →          Σ[ φ ∈ (hom 𝑪 𝑩)] ∣ τ ∣ ≡ ∣ φ ∣ ∘ ∣ ν ∣
+
+ HomFactor fxy wd τ ν Kντ νE = (φ , φIsHomCB) , τφν
+  where
+   νInv : C → A
+   νInv = SurjInv ∣ ν ∣ νE
+
+   η : ∀ c → ∣ ν ∣ (νInv c) ≡ c
+   η c = SurjInvIsRightInv ∣ ν ∣ νE c
+
+   φ : C → B
+   φ = ∣ τ ∣ ∘ νInv
+
+   ξ : ∀ a → kernel ∣ ν ∣ (a , νInv (∣ ν ∣ a))
+   ξ a = (η (∣ ν ∣ a))⁻¹
+
+   τφν : ∣ τ ∣ ≡ φ ∘ ∣ ν ∣
+   τφν = fxy λ x → Kντ (ξ x)
+
+   φIsHomCB : ∀ 𝑓 c → φ ((𝑓 ̂ 𝑪) c) ≡ ((𝑓 ̂ 𝑩)(φ ∘ c))
+   φIsHomCB 𝑓 c = φ ((𝑓 ̂ 𝑪) c)     ≡⟨ PE.cong φ (wd (𝑓 ̂ 𝑪) c (∣ ν ∣ ∘ (νInv ∘ c)) (λ i → (η (c i))⁻¹)) ⟩
+                  φ ((𝑓 ̂ 𝑪)(∣ ν ∣ ∘(νInv ∘ c)))   ≡⟨ PE.cong φ (∥ ν ∥ 𝑓 (νInv ∘ c))⁻¹ ⟩
+                  φ (∣ ν ∣((𝑓 ̂ 𝑨)(νInv ∘ c)))     ≡⟨ cong-app(τφν ⁻¹)((𝑓 ̂ 𝑨)(νInv ∘ c))⟩
+                  ∣ τ ∣((𝑓 ̂ 𝑨)(νInv ∘ c))         ≡⟨ ∥ τ ∥ 𝑓 (νInv ∘ c) ⟩
+                  (𝑓 ̂ 𝑩)(λ x → ∣ τ ∣(νInv (c x))) ∎
+
+\end{code}
+
+If, in addition to the hypotheses of the last theorem, we assume τ is epic, then so is φ. (Note that the proof also requires an additional local function extensionality postulate, `funext β β`.)
+
+\begin{code}
+
+ HomFactorEpi : funext α β → swelldef 𝓥 γ
```

Comment:
Same here.

Reply:
Removed funext from the premises.  (Will update the PR now.)

Reply:
I saw the other updates, but not this one?

Reply:
Hmmm... I'm not sure why you didn't see this update.  Maybe because it's from a follow-up/add-on PR.